include/boost/url/any_allocator.hpp

```diff
@@ -0,0 +1,170 @@
+//
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+//
+// Official repository: https://github.com/CPPAlliance/url
+//
+
+#ifndef BOOST_URL_ANY_ALLOCATOR_HPP
+#define BOOST_URL_ANY_ALLOCATOR_HPP
+
+#include <boost/url/detail/config.hpp>
+
+#include <memory>
+#include <cstddef>
+#include <type_traits>
+#include <typeindex>
+#include <atomic>
+
+namespace boost {
+namespace urls {
+
+// move to another header
+namespace detail {
+    struct any_allocator_base {
+        std::type_index type{typeid(void)};
+
+        virtual
+        ~any_allocator_base() {}
+
+        explicit
+        any_allocator_base(std::type_index t) : type(t) {}
+
+        virtual
+        char *
+        allocate(std::size_t n, std::size_t size, std::size_t align) = 0;
+
+        virtual
+        void
+        deallocate(char *p, std::size_t n, std::size_t size, std::size_t align) = 0;
+
+        virtual
+        bool
+        is_equal(const any_allocator_base &p) = 0;
+
+    };
+}
+
+/** A type-erased allocator with shared ownership.
+ */
+template <class T>
+class any_allocator {
+    std::shared_ptr<detail::any_allocator_base> p_ = nullptr;
+
+    template <class U>
+    friend class any_allocator;
+
+public:
+    /** The type any_allocator allocates
+    */
+    using value_type = T;
+
+    /** Determine the type of any_allocator for type U
+    */
+    template<class U>
+    struct rebind
+    {
+        /** An any_allocator for type U
+         */
+        using other = any_allocator<U>;
+    };
+
+    /** Determine with two allocator instances are always equal
+    */
+    using is_always_equal = std::false_type;
+
+    /** Constructor
+    */
+    any_allocator() = default;
+
+    /** Constructor
+    */
+    any_allocator(
+        any_allocator const& other) noexcept;
+
+    /** Constructor
+    */
+    any_allocator(
+        any_allocator&& other) noexcept;
+
+    /** Constructor
+    */
+    template <class U>
+    explicit
+    any_allocator(
+        any_allocator<U> const& other) noexcept;
+
+    /** Constructor
+    */
+    template <class U>
+    explicit
+    any_allocator(
+        any_allocator<U> && other) noexcept;
+
+    /** Copy assignment
+    */
+    any_allocator&
+    operator=(
+        any_allocator const& other) noexcept;
+
+    /** Constructor
+    */
+    template<class Allocator>
+    explicit
+    any_allocator(Allocator const& a) noexcept;
+
+    /** Allocates uninitialized storage
+    */
+    T*
+    allocate(
+        std::size_t n) const;
+
+    /** Deallocates storage
+    */
+    void
+    deallocate(
+        T* p,
+        std::size_t n);
+
+    /** Compares two allocator instances
+    */
+    friend
+    bool
+    operator==(
+        const any_allocator& a,
+        const any_allocator& b) noexcept
+    {
+        if (a.p_ == b.p_) {
+            return true;
+        } else if (a.p_ == nullptr || b.p_ == nullptr) {
```

Comment:
no no.. default-constructed `any_allocator` needs to work correctly. Probably pointing to an instance of `std::allocator<char>` with static storage duration.

Reply:
Ah I see later on that you are checking for an empty shared_ptr.